### ISSUE TITLE
Neovide: Build against amdvlk

### ISSUE THREAD

**Issue by enderger** [NixOS/nixpkgs#123262]
**Describe the bug**
Due to an issue with Neovide, some computers need to use the `amdvlk` driver instead of the standard version. The package works on non-AMD hardware and usually fixes the issue.

**To Reproduce**
Unfortunately, the issue is inconsistent, however on a number of machines attempting to run Neovide causes the crash. It's a known issue.

**Screenshots**
Error message:
```
Ignored client type property: "methods"
Ignored client type property: "attributes"
thread 'main' panicked at 'Failed to create renderer: CreateInstanceError(InstanceError(VkError(ERROR_EXTENSION_NOT_PRESENT)))', src/window/sdl2/mod.rs:456:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Additional context**
See the Neovide install section.

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
> - system: `"x86_64-linux"`
> - host os: `Linux 5.11.16-300.fc34.x86_64, Fedora, 34 (Workstation Edition)`
> - multi-user?: `no`
> - sandbox: `yes`
> - version: `nix-env (Nix) 2.4pre20210317_8a5203d`
> - channels(daniel): `"nixpkgs-21.05pre278688.c0e88185200"`
> - nixpkgs: `/home/daniel/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: neovide
# a list of nixos modules affected by the problem
module:
```

**Comment by teto** [comment]
I wanted to try neovide on another computer with a dedicated GPU but I get on master:
```
  running: "/nix/store/jabh25wi8r7v2gpkfyhvm9jaxpzgw73x-gcc-wrapper-10.3.0/bin/c++" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/nix/store/3dwvn3b2wf5ydhrqs0kkzxppzfxig9ih-source" "-Wall" "-Wextra" "-std=c++17" "-DNDEBUG=" "-DSK_R32_SHIFT=16" "-DSK_GAMMA_APPLY_TO_A8=" "-DSK_USE_VMA=" "-DSKIA_IMPLEMENTATION=1" "-DSK_VULKAN=" "-DSK_SUPPORT_PDF=" "-DSK_CODEC_DECODES_JPEG=" "-DSK_ENCODE_JPEG=" "-DSK_USE_LIBGIFCODEC=" "-DSK_CODEC_DECODES_PNG=" "-DSK_ENCODE_PNG=" "-DSK_XML=" "-DSKSHAPER_IMPLEMENTATION=1" "-DSKUNICODE_IMPLEMENTATION=1" "-DU_SHOW_CPLUSPLUS_API=0" "-DSK_SHAPER_HARFBUZZ_AVAILABLE=" "-DSK_UNICODE_AVAILABLE=" "-DU_USING_ICU_NAMESPACE=0" "-DU_DISABLE_RENAMING=" "-DSK_USING_THIRD_PARTY_ICU=" "-o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/paragraph.o" "-c" "src/paragraph.cpp"
  exit code: 0
  running: "/nix/store/jabh25wi8r7v2gpkfyhvm9jaxpzgw73x-gcc-wrapper-10.3.0/bin/c++" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/nix/store/3dwvn3b2wf5ydhrqs0kkzxppzfxig9ih-source" "-Wall" "-Wextra" "-std=c++17" "-DNDEBUG=" "-DSK_R32_SHIFT=16" "-DSK_GAMMA_APPLY_TO_A8=" "-DSK_USE_VMA=" "-DSKIA_IMPLEMENTATION=1" "-DSK_VULKAN=" "-DSK_SUPPORT_PDF=" "-DSK_CODEC_DECODES_JPEG=" "-DSK_ENCODE_JPEG=" "-DSK_USE_LIBGIFCODEC=" "-DSK_CODEC_DECODES_PNG=" "-DSK_ENCODE_PNG=" "-DSK_XML=" "-DSKSHAPER_IMPLEMENTATION=1" "-DSKUNICODE_IMPLEMENTATION=1" "-DU_SHOW_CPLUSPLUS_API=0" "-DSK_SHAPER_HARFBUZZ_AVAILABLE=" "-DSK_UNICODE_AVAILABLE=" "-DU_USING_ICU_NAMESPACE=0" "-DU_DISABLE_RENAMING=" "-DSK_USING_THIRD_PARTY_ICU=" "-o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/svg.o" "-c" "src/svg.cpp"
  exit code: 0
  running: "ar" "cq" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/libskia-bindings.a" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/bindings.o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/vulkan.o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/gpu.o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/shaper.o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/paragraph.o" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/src/svg.o"
  exit code: 0
  running: "ar" "s" "/build/source/target/x86_64-unknown-linux-gnu/release/build/skia-bindings-c6613dc34168c33d/out/skia/libskia-bindings.a"
  exit code: 0
  GENERATING BINDINGS

  --- stderr
  thread 'main' panicked at 'Unable to find libclang: "couldn\'t find any valid shared libraries matching: [\'libclang.so\', \'libclang-*.so\', \'libclang.so.*\', \'libclang-*.so.*\'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"', /build/neovide-20210515-vendor.tar.gz/bindgen/src/lib.rs:1896:31
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  ```

  Also I wonder if anyone tried the opengl branch ?

**Comment by teto** [comment]
cc @ck3d can you reproduce my issue ?

**Comment by ck3d** [comment]
Yes, it was caused by merging stage branch. I created two PRs, one for release-21.05 and one for master:

- https://github.com/NixOS/nixpkgs/pull/124140
- https://github.com/NixOS/nixpkgs/pull/124141

**Comment by ck3d** [comment]
regarding the main issue:
I use neovide on AMD GPU with video driver amdgpu which is based on mesas vulkan implementation. I did not install amdvkl. Which video driver do you use?

**Comment by enderger** [comment]
On my Fedora system, I use the Intel drivers (though upon further analysis, Vulkan packages in general don't work well through Nix on it). On my NixOS box (whose error message is simply "could not create surface"), I use the `radv` drivers.

**Comment by ck3d** [comment]
radv is the default driver which I also use. I tried amdvlk and had also no issues.
Does the application vkcube from package vulkan-tools run on your system?

**Comment by gebner** [comment]
@enderger For Fedora, you probably need to use nixGL: https://github.com/guibou/nixGL

I get the same error with an RX 480, no matter whether I use amdvlk or radv:
```
Ignored client type property: "methods"
Ignored client type property: "attributes"
thread 'main' panicked at 'Could not create surface', src/renderer/rendered_window.rs:40:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by gebner** [comment]
Apparently we're not the only ones with issues: https://github.com/NixOS/nixpkgs/pull/122998#issuecomment-841379718

**Comment by enderger** [comment]
Ok, your fix worked on the Fedora machine. However, the NixOS box wouldn't be affected by this. I think that Neovide issue [#459](https://github.com/Kethku/neovide/issues/459) is the culprit for it's issue.

**Comment by enderger** [comment]
Here's my NixOS box's `vulkaninfo` output:
```                                                                                                             
$  vulkaninfo | rg "GPU id :"
WARNING: lavapipe is not a conformant vulkan implementation, testing use only.
GPU id : 0 (AMD RADV POLARIS12 (ACO)):
GPU id : 1 (llvmpipe (LLVM 11.1.0, 256 bits)):
```

**Comment by teto** [comment]
I managed to start neovide using the opengl branch and patching manually the program
```
  neovide = prev.neovide.overrideAttrs(oa:
    let
      rpathLibs = with final.pkgs; [
        libglvnd
        freeglut
        freeglut.dev
        freetype

        vulkan-loader
        xorg.libXcursor
        xorg.libXext
        xorg.libXrandr
        xorg.libXi
        fontconfig
      ];
    in {
    # shellHook = ''echo "hello world"'';
      buildInputs = with final.pkgs; oa.buildInputs ++ [
        libglvnd
        freeglut
        freeglut.dev
        freetype

        vulkan-loader
        xorg.libXcursor
        xorg.libXext
        xorg.libXrandr
        xorg.libXi
        fontconfig
      ];

      shellHook = ''
        echo "run this after build"
        echo 'patchelf --set-rpath "${final.lib.makeLibraryPath rpathLibs}" target/debug/neovide'
      '';
  });
```
Haven't tried to clean it up yet.

**Comment by bnjmnt4n** [comment]
This issue can probably be closed since #127506 was merged.
